### PR TITLE
Use the ID as primary key for User equality checking

### LINKED ISSUE #5631: Failure in test if an entity has a oneToMany relationship to User (with DTO and ES)
##### **Overview of the issue**

The generated createEntity test fail with the following stacktrace :
<details>

```sh
org.springframework.data.elasticsearch.ElasticsearchException: failed to map source [ {"id":1,"userOneToMany":{"id":5,"login":null,"firstName":null,"lastName":null,"email":null,"activated":false,"langKey":null,"imageUrl":null,"resetKey":null,"resetDate":null},"userManyToManies":[],"userOneToOne":null}] to class TestEntity

	at org.springframework.data.elasticsearch.core.AbstractResultMapper.mapEntity(AbstractResultMapper.java:42)
	at org.springframework.data.elasticsearch.core.DefaultResultMapper.mapResult(DefaultResultMapper.java:154)
	at org.springframework.data.elasticsearch.core.ElasticsearchTemplate.queryForObject(ElasticsearchTemplate.java:251)
	at org.springframework.data.elasticsearch.core.ElasticsearchTemplate.queryForObject(ElasticsearchTemplate.java:241)
	at org.springframework.data.elasticsearch.repository.support.AbstractElasticsearchRepository.findOne(AbstractElasticsearchRepository.java:100)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.executeMethodOn(RepositoryFactorySupport.java:504)
	at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.doInvoke(RepositoryFactorySupport.java:489)
	at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.invoke(RepositoryFactorySupport.java:461)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.data.projection.DefaultMethodInvokingMethodInterceptor.invoke(DefaultMethodInvokingMethodInterceptor.java:61)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:92)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.data.repository.core.support.SurroundingTransactionDetectorMethodInterceptor.invoke(SurroundingTransactionDetectorMethodInterceptor.java:57)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:213)
	at com.sun.proxy.$Proxy201.findOne(Unknown Source)
	at com.mycompany.myapp.grpc.entity.test_entity.TestEntityGrpcServiceIntTest.createTestEntity(TestEntityGrpcServiceIntTest.java:121)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.springframework.test.context.junit4.statements.RunBeforeTestMethodCallbacks.evaluate(RunBeforeTestMethodCallbacks.java:75)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.springframework.test.context.junit4.statements.RunAfterTestMethodCallbacks.evaluate(RunAfterTestMethodCallbacks.java:86)
	at org.springframework.test.context.junit4.statements.SpringRepeat.evaluate(SpringRepeat.java:84)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:252)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:94)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.springframework.test.context.junit4.statements.RunBeforeTestClassCallbacks.evaluate(RunBeforeTestClassCallbacks.java:61)
	at org.springframework.test.context.junit4.statements.RunAfterTestClassCallbacks.evaluate(RunAfterTestClassCallbacks.java:70)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.run(SpringJUnit4ClassRunner.java:191)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:51)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:237)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
Caused by: com.fasterxml.jackson.databind.JsonMappingException: N/A
 at [Source: {"id":1,"userOneToMany":{"id":5,"login":null,"firstName":null,"lastName":null,"email":null,"activated":false,"langKey":null,"imageUrl":null,"resetKey":null,"resetDate":null},"userManyToManies":[],"userOneToOne":null}; line: 1, column: 41] (through reference chain: com.mycompany.myapp.domain.TestEntity["userOneToMany"]->com.mycompany.myapp.domain.User["login"])
	at com.fasterxml.jackson.databind.JsonMappingException.from(JsonMappingException.java:277)
	at com.fasterxml.jackson.databind.deser.SettableBeanProperty._throwAsIOE(SettableBeanProperty.java:551)
	at com.fasterxml.jackson.databind.deser.SettableBeanProperty._throwAsIOE(SettableBeanProperty.java:532)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:105)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:357)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:148)
	at com.fasterxml.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:504)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:101)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:357)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:148)
	at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:3798)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:2842)
	at com.mycompany.myapp.config.ElasticsearchConfiguration$CustomEntityMapper.mapToObject(ElasticsearchConfiguration.java:40)
	at org.springframework.data.elasticsearch.core.AbstractResultMapper.mapEntity(AbstractResultMapper.java:40)
	... 51 more
Caused by: java.lang.NullPointerException
	at com.mycompany.myapp.domain.User.setLogin(User.java:115)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:103)
	... 61 more
```
</details>

##### **Reproduce the error**
I ran into it when setting dto to all travis relationship test entities (Test*.json). See https://github.com/cbornet/generator-jhipster/commit/f98ca7f66eec1512f3975e1616aca9d72a30760e
Then [this build](https://travis-ci.org/cbornet/generator-jhipster/jobs/223608965) failed in Travis

##### **Suggest a Fix**
The crash occurs in createTestEntity() when getting the entity from ES (```TestEntity testEntityEs = testEntitySearchRepository.findOne(testTestEntity.getId());```)
This is because the user login is null in ES so user.setLogin(null) is called but the code is
```java
    //Lowercase the login before saving it in database
    public void setLogin(String login) {
        this.login = login.toLowerCase(Locale.ENGLISH);
    }
```
hence NPE ! 💥 

Adding a null-check here makes things a little better but the test will still not pass because the user equality that is done just after will fail since it's done on the login.
Going a little backward in time, we can see that the login was not set in ES because since we're using DTOs, only the ID was mapped when creating the userOneToMany (calling `userMapper.userFromId(testEntityDTO.getUserOneToManyId())`)
```java
    @Mapping(target = "testManyToOnes", ignore = true)
    @Mapping(target = "testManyToManies", ignore = true)
    @Mapping(target = "testOneToOne", ignore = true)
    @Mapping(source = "userOneToManyId", target = "userOneToMany")
    @Mapping(source = "userOneToOneId", target = "userOneToOne")
    @Mapping(target = "testCustomTableNames", ignore = true)
    TestEntity testEntityDTOToTestEntity(TestEntityDTO testEntityDTO);
```

From there I'm not sure what should be the fix.
* Do we expect the user login to be saved in ES together with the ID (since it's the business key) ? Then we will have to see how to do that cleanly with mapstruct.
* Or is the ID sufficient to identify the user ? Then if the ID is the true identifier of a user, shouldn't it be used to check equality ?

- [x] Checking this box is mandatory (this is just to show you read everything)

### PR DESCRIPTION
**Pull request by cbornet**
Fix #5631

- Please make sure the below checklist is followed for Pull Requests.

- [x] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
- [x] Tests are added where necessary
- [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

File: generators/server/templates/src/main/java/package/domain/_User.java [hunk 1]
```diff
@@ -183,7 +184,7 @@ public String getLogin() {
 
     //Lowercase the login before saving it in database
     public void setLogin(String login) {
-        this.login = login.toLowerCase(Locale.ENGLISH);
+        this.login = login == null ? null : login.toLowerCase(Locale.ENGLISH);
```
Comment (pascalgrimaud) [inline generators/server/templates/src/main/java/package/domain/_User.java]: @cbornet : for your information, Sonar complains about this line [here](https://sonarcloud.io/dashboard?id=io.github.jhipster.sample%3Ajhipster-sample-application%3Asrc%2Fmain%2Fjava%2Fio%2Fgithub%2Fjhipster%2Fsample%2Fdomain%2FUser.java)
> "login" is marked "javax.validation.constraints.NotNull" but is set to null.

Comment (jdubois) [inline generators/server/templates/src/main/java/package/domain/_User.java]: yes, this makes our Sonar build fail :-(
I also don't think this code is easy to read

Comment (deepu105) [inline generators/server/templates/src/main/java/package/domain/_User.java]: Oops

Comment (cbornet) [inline generators/server/templates/src/main/java/package/domain/_User.java]: Ok I'm on it

**Pull request merged**